File: apps/web/lib/api/subtitle-gen.ts
import { request } from "@/lib/http";

/** 后端统一响应信封（见 movieclaw_api.schemas.response.ApiResponse） */
interface ApiEnvelope<T> {
  success: boolean;
  code: string;
  message: string;
  data: T;
}

async function unwrap<T>(promise: Promise<ApiEnvelope<T>>): Promise<T> {
  return (await promise).data;
}

/**
 * AI 字幕生成（docs/design/subtitle-ai-translate.md）：选源 → 同步质检 →
 * LLM 翻译 → 机检 → 落盘外挂 sidecar。管理员专属（消费 LLM 配额）。
 */

/** 一个候选参考字幕的打分结论（预检确认框展示）。 */
export interface SubgenCandidate {
  kind: "embedded" | "external";
  key: string;
  language: string | null;
  format: string | null;
  /** 非空 = 被排除原因（forced/目标语言/图形轨） */
  excluded: string | null;
  /** 打分明细："为什么选了这条"要能看懂 */
  reasons: string[];
}

export interface SubgenPreview {
  candidates: SubgenCandidate[];
  /** "kind:key"；null = 没有可用参考 */
  chosen_key: string | null;
  event_count: number;
  estimated_tokens: number;
  /** 目标语言 AI 字幕已存在（再生成 = 覆盖） */
  already_generated: boolean;
  warnings: string[];
}

export interface SubgenQuality {
  event_count: number;
  cps_overrun: number;
  overlong: number;
  kept_original: number;
  glossary_usage: Record<string, number>;
}

export interface SubgenResult {
  ok: boolean;
  message: string;
  filename: string | null;
  sync_score: number | null;
  source_desc: string | null;
  report: SubgenQuality | null;
  finished_at: string | null;
}

export interface SubgenProgress {
  running: boolean;
  phase: string | null;
  message: string | null;
  done_blocks: number;
  total_blocks: number;
  last_result: SubgenResult | null;
}

export interface CalibrateResult {
  ok: boolean;
  message: string;
  scale: number | null;
  offset_ms: number | null;
  score: number | null;
}

/** 生成预检：选源结果 + 成本估算（确认框素材，不动 LLM）。 */
export function subgenPreview(fileId: number, targetLanguage = "chs"): Promise<SubgenPreview> {
  const query = new URLSearchParams({ target_language: targetLanguage });
  return unwrap(
    request<ApiEnvelope<SubgenPreview>>(
      `/library/files/${fileId}/subtitles/generate/preview?${query}`,
    ),
  );
}

/** 发起生成（后台执行；同文件已在跑时后端 400）。 */
export function subgenStart(
  fileId: number,
  targetLanguage = "chs",
): Promise<{ preview: SubgenPreview }> {
  return unwrap(
    request<ApiEnvelope<{ preview: SubgenPreview }>>(
      `/library/files/${fileId}/subtitles/generate`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ target_language: targetLanguage }),
      },
    ),
  );
}

/** 任务状态：进行中带块进度，否则带最近一次结论。 */
export function subgenStatus(fileId: number): Promise<SubgenProgress> {
  return unwrap(
    request<ApiEnvelope<SubgenProgress>>(`/library/files/${fileId}/subtitles/generate/status`),
  );
}

/** 停止进行中的生成（下一块边界生效；已译块保留可续传）。 */
export function subgenStop(fileId: number): Promise<{ stopped: boolean }> {
  return unwrap(
    request<ApiEnvelope<{ stopped: boolean }>>(
      `/library/files/${fileId}/subtitles/generate/stop`,
      { method: "POST" },
    ),
  );
}

/** 一键校准外挂字幕时间轴（零 LLM 成本；校准后覆盖写回）。 */
export function calibrateSubtitle(fileId: number, filename: string): Promise<CalibrateResult> {
  return unwrap(
    request<ApiEnvelope<CalibrateResult>>(`/library/files/${fileId}/subtitles/calibrate`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ filename }),
    }),
  );
}

File: src/movieclaw_api/schemas/subtitle_gen.py
"""AI 字幕生成接口的响应形态（docs/design/subtitle-ai-translate.md §6）。"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class SourceCandidateView(BaseModel):
    kind: str  # embedded/external
    key: str
    language: str | None
    format: str | None
    excluded: str | None  # 非空 = 被排除原因
    reasons: list[str]  # 打分明细（用户要能看懂"为什么选了这条"）


class GenPreviewView(BaseModel):
    """发起前的确认素材：选源结果 + 成本估算。"""

    candidates: list[SourceCandidateView]
    chosen_key: str | None  # kind:key；None=无可用参考
    event_count: int
    estimated_tokens: int
    already_generated: bool  # 目标语言的 AI 字幕已存在（再生成=覆盖）
    warnings: list[str]


class GenQualityView(BaseModel):
    """机检报告（§3.5）。"""

    event_count: int
    cps_overrun: int
    overlong: int
    kept_original: int
    glossary_usage: dict[str, int]


class GenResultView(BaseModel):
    ok: bool
    message: str
    filename: str | None
    sync_score: float | None
    source_desc: str | None
    report: GenQualityView | None
    finished_at: datetime | None


class GenProgressView(BaseModel):
    """任务状态查询：running 时带实时进度，否则带最近一次结论。"""

    running: bool
    phase: str | None = None
    message: str | None = None
    done_blocks: int = 0
    total_blocks: int = 0
    last_result: GenResultView | None = None


class GenStartPayload(BaseModel):
    target_language: str = "chs"  # 文件名语言 token（默认简体中文）


class GenStartView(BaseModel):
    """已入队的确认回执（含预估，前端展示）。"""

    preview: GenPreviewView


class CalibratePayload(BaseModel):
    filename: str  # 台账里的外挂字幕文件名


class CalibrateResultView(BaseModel):
    ok: bool
    message: str
    scale: float | None
    offset_ms: int | None
    score: float | None
